tutorials/todo/src/main.sw

```diff
@@ -0,0 +1,160 @@
+            get::<Todo>(TODO_KEY_2),
+            get::<Todo>(TODO_KEY_3),
+            get::<Todo>(TODO_KEY_4),
+        ]
+    }
+    fn toggle_todo(index: u8) -> [Todo; 5] {
+        // @see TODO: ^
+        match index {
+            0 => {
+                let current = get::<Todo>(TODO_KEY_0);
+                store(TODO_KEY_0, Todo {
+                    value: current.value,
+                    completed: !current.completed,
+                });
+            },
+            1 => {
+                let current = get::<Todo>(TODO_KEY_1);
+                store(TODO_KEY_1, Todo {
+                    value: current.value,
+                    completed: !current.completed,
+                });
+            },
+            2 => {
+                let current = get::<Todo>(TODO_KEY_2);
+                store(TODO_KEY_2, Todo {
+                    value: current.value,
+                    completed: !current.completed,
+                });
+            },
+            3 => {
+                let current = get::<Todo>(TODO_KEY_3);
+                store(TODO_KEY_3, Todo {
+                    value: current.value,
+                    completed: !current.completed,
+                });
+            },
+            4 => {
+                let current = get::<Todo>(TODO_KEY_4);
+                store(TODO_KEY_4, Todo {
+                    value: current.value,
+                    completed: !current.completed,
+                });
+            },
+            _ => {
+                
+            },
+        };
+
+        // @see TODO: ^
+        [
+            get::<Todo>(TODO_KEY_0),
+            get::<Todo>(TODO_KEY_1),
+            get::<Todo>(TODO_KEY_2),
+            get::<Todo>(TODO_KEY_3),
+            get::<Todo>(TODO_KEY_4),
+        ]
+    }
+   fn remove_todo(index: u8) -> [Todo; 5] {
+        let value: str[20] = "                    ";
+        let todo = Todo {
+            value,
+            completed: false,
+        };
+        // @see TODO: ^
+        match index {
+            0 => {
+                store(TODO_KEY_0, todo);
+            },
+            1 => {
+                store(TODO_KEY_1, todo);
+            },
+            2 => {
+                store(TODO_KEY_2, todo);
+            },
+            3 => {
+                store(TODO_KEY_3, todo);
+            },
+            4 => {
+                store(TODO_KEY_4, todo);
+            },
+            _ => {
+                
+            },
+        };
+
+        // @see TODO: ^
+        [
+            get::<Todo>(TODO_KEY_0),
+            get::<Todo>(TODO_KEY_1),
+            get::<Todo>(TODO_KEY_2),
+            get::<Todo>(TODO_KEY_3),
+            get::<Todo>(TODO_KEY_4),
+        ]
+    }
+    fn get_todos() -> [Todo; 5] {
+        [
+            get::<Todo>(TODO_KEY_0),
+            get::<Todo>(TODO_KEY_1),
+            get::<Todo>(TODO_KEY_2),
+            get::<Todo>(TODO_KEY_3),
+            get::<Todo>(TODO_KEY_4),
+        ]
+    }
+}
```

Comment:
```suggestion
contract;

dep abi;
dep tasks;

use std::storage::StorageMap;
use abi::Todo;
use tasks::Task;

// move to `task.sw` and declare as a library
struct Task {
    completed: bool,
    value: str[20],
}

// move to `abi.sw`, declare it as a library and add "dep task;" followed by "use task::Task;"
abi Todo {
    fn get_tasks() -> [Task; 5];
    fn add_task(index: u64, value: str[20]);
    fn toggle_task(index: u64);
    fn remove_task(index: u64);
}

storage {
    tasks: StorageMap<u64, Task>,
}

// TODO: remove hack array once mutable arrays are ready see https://github.com/FuelLabs/sway/issues/1626

impl Todo for Contract {
     fn add_task(index: u64 value: str[20]) {
        // logic needs to be fixed to make sure they can only add up to 5 tasks
        storage.tasks.insert(index, Task {
            value,
            completed: false,
        });
    }
    
    fn toggle_task(index: u64) {
        // logic needs to be fixed to make sure they only toggle a task that exists
        let mut task = strorage.tasks.get(index);
        task.completed = !task.completed;
        storage.tasks.insert(index, task);
    }
    
   fn remove_task(index: u64) {
        // Note: that this doesn't actually remove anything... it actually adds a new task at `index`
        // this should probably be fixed with better logic for an example
        storage.tasks.insert(index, Task {
            value: "                    ",
            completed: false,
        });
    }
    
    fn get_tasks() -> [Task; 5] {
        // logic needs to be fixed since the user can add a task at any index
        [
            storage.tasks.get(0),
            storage.tasks.get(1),
            storage.tasks.get(2),
            storage.tasks.get(3),
            storage.tasks.get(4),
        ]
    }
}
```

This needs some improvements but this is a general idea that you should be able to alter to use proper error handling